src/Http/WebUtilities/src/MultipartSectionPipeReader.cs

```diff
@@ -0,0 +1,360 @@
+            {
+                throw new IOException("Unexpected end of Stream, the content may have already been read by another component. ");
+            }
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> body, _boundary.BoundaryBytes[_partialMatchIndex]))
+                {
+                    //advance reader to end
+                    sequence = sequence.Slice(sequence.End);
+                    return (false, sequenceReader.Length);
+                }
+
+                read += body.Length;
+
+                if (_partialMatchIndex != 0 && body.Length > 0)
+                {
+                    // there was a potential end, but it actually isn't
+                    read += _partialMatchIndex; //include *previous* matches as read
+                    _partialMatchIndex = 0;
+                    sequence = sequence.Slice(sequenceReader.Consumed - 1); // the current _partialMatchIndex might be a begining to a boundary
+                    return (false, read);
+                }
+
+                //continue to check for boundaryMatch
+                for (int i = _partialMatchIndex + 1; i < _boundary.BoundaryBytes.Length; i++)
+                {
+                    if (sequenceReader.TryRead(out var value))
+                    {
+                        if (_boundary.BoundaryBytes[i] == value)
+                        {
+                            continue;
+                        }
+
+                        //no match - mark previous matches as read, the current one might be the begining of a new boundary
+                        read += i;
+                        sequence = sequence.Slice(sequenceReader.Consumed - 1);
+                        return (false, read);
+
+                    }
+
+                    //end of sequence
+                    //There still might be a partial Match, we need to read more to know for sure
+                    _partialMatchIndex = i;
+                    sequence = sequence.Slice(read);
+                    //also slice partial read to insure _bufferedData is empty in the next check
+                    sequence = sequence.Slice(i);
+                    return (false, read);
+                }
+
+                //boundary was found
+                sequence = sequence.Slice(sequenceReader.Position);
+                RawLength += _boundary.BoundaryBytes.Length;
+                return (true, read);
+            }
+            return (false, read);
+        }
+
+        private bool TrySkipMetadata(ref ReadOnlySequence<byte> sequence, bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(sequence);
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out var body, CrlfDelimiter))
+                {
+                    if (sequenceReader.TryReadTo(out var _, EndOfFileDelimiter))
```

Comment:
@Tratcher not sure how to put this in a reply, but It references `CrlfDelimeter` and `EndOfFileDelimeter` right here.